### ISSUE TITLE
Migrate from MongoDB to MariaDB

### ISSUE THREAD

**Issue by nzalev** [Ylianst/MeshCentral#2423]
I was not able to find anything in the documentation or other issues. Is there a supported way to migrate from an existing MongoDB database to MariaDB?

I understand I can manually do it as well by getting the data from mongodb and creating queries to insert, but this is something I want to avoid if possible.

**Comment by Ylianst** [comment]
You can try these commands:

```
node node_modules/meshcentral --dbexport server.json
node node_modules/meshcentral --dbimport server.json
```

Stop the server, export the db into JSON, change the database in the config.json and import the JSON again. This will migrate the devices, device groups, accounts user groups, user notes, etc. It will not migrate power state and events.

Let me know if that works,
Ylian

**Comment by nzalev** [comment]
> You can try these commands:
> 
> ```
> node node_modules/meshcentral --dbexport server.json
> node node_modules/meshcentral --dbimport server.json
> ```
> 
> Stop the server, export the db into JSON, change the database in the config.json and import the JSON again. This will migrate the devices, device groups, accounts user groups, user notes, etc. It will not migrate power state and events.
> 
> Let me know if that works,
> Ylian

Interestingly, it did not import all the data. (or it at least does not display all the imported data)

The export and import both listed the same number of objects, however, half the device groups and associated agents are missing on the imported/mariadb server.

I am just starting to look at the data in the MariaDB main table.

**Comment by nzalev** [comment]
@Ylianst I have tracked down the issue.

The problem is not to do with the device groups/meshes. The issue is actually with links in the user, in my case.
An example of a problem link:
```
    'mesh//88CO9BnfLhKxo%246sibFlR3zMysxjGXLmBzN7OJN7vZRWepkk2wnek4OLAOUHdDqw': { rights: 4294967295 }
```

After you run `node node_modules/meshcentral --dbimport server.json`, the database code will eventually reach
```javascript
        obj.Set = function (value, func) {
            obj.dbCounters.fileSet++;
            var extra = null, extraex = null;
            value = common.escapeLinksFieldNameEx(value);
            if (value.meshid) { extra = value.meshid; } else if (value.email) { extra = 'email/' + value.email; } else if (value.nodeid) { extra = value.nodeid; }
            if ((value.type == 'node') && (value.intelamt != null) && (value.intelamt.uuid != null)) { extraex = 'uuid/' + value.intelamt.uuid; }
            if (value._id == null) { value._id = require('crypto').randomBytes(16).toString('hex'); }
            sqlDbQuery('REPLACE INTO meshcentral.main VALUE (?, ?, ?, ?, ?, ?)', [value._id, (value.type ? value.type : null), ((value.domain != null) ? value.domain : null), extra, extraex, JSON.stringify(performTypedRecordEncrypt(value))], func);
        }
```

The assignment `value = common.escapeLinksFieldNameEx(value);` transforms the link in a harmful way (harmful in this case) by replacing `%` with `%25` (return name.split('%').join('%25'))
```
Before:
    'mesh//88CO9BnfLhKxo%246sibFlR3zMysxjGXLmBzN7OJN7vZRWepkk2wnek4OLAOUHdDqw': { rights: 4294967295 }
    'mesh//88CO9BnfLhKxo%25246sibFlR3zMysxjGXLmBzN7OJN7vZRWepkk2wnek4OLAOUHdDqw': { rights: 4294967295 }
After ^
```

This will be inserted into MariaDB as-is.

Now, when the webserver starts and it performs user link cleanup, it will remove these affected links.

Adding a `cosole.log(user.links[j])` before the deletion in
```javascript
    else if (j.startsWith('mesh/') && ((obj.meshes[j] == null) || (obj.meshes[j].deleted != null))) { delete user.links[j]; } 
```
will spit out `{ rights: 4294967295 }`

**Comment by Ylianst** [comment]
Ohhhh wow!!! You got it. You can tell I don't use MariaDB much. I can work on this tonight.

**Comment by Ylianst** [comment]
Ok. just got done with #2424. Working on this one now. I happen to have a MongoDB database I can export and import into MariaDB.

**Comment by Ylianst** [comment]
I see, the links in the exported file are already "escaped". So in this case:

```
'mesh//88CO9BnfLhKxo%246sibFlR3zMysxjGXLmBzN7OJN7vZRWepkk2wnek4OLAOUHdDqw'
```

The %24 is already an escaped character... but importing it double-escapes it... which is bad.

**Comment by nzalev** [comment]
Great. Thank you.

**Comment by Ylianst** [comment]
Just published MeshCentral v0.8.0 with a fix. The main issue is that InsertMany() should not be doing any escaping when inserting, but it did for MariaDB and MySQL. Latest version should fix this. Let me know if it works.

**Comment by nzalev** [comment]
> Just published MeshCentral v0.8.0 with a fix. The main issue is that InsertMany() should not be doing any escaping when inserting, but it did for MariaDB and MySQL. Latest version should fix this. Let me know if it works.

I installed v0.8.0 and the db import looks to be working well. Thanks.
